tensorflow/compiler/xla/service/algebraic_simplifier.cc

```diff
@@ -1665,6 +1668,228 @@ StatusOr<HloInstruction*> AlgebraicSimplifierVisitor::OptimizeDotOfGather(
+// dimensions on both sides of the dot in the same way, the result of the
+// dot is not affected.
+StatusOr<HloInstruction*>
+AlgebraicSimplifierVisitor::OptimizeDotOfReorderContractingDims(
+    HloInstruction* dot) {
+  // This transformation assumes layout is not assigned yet.
+  if (options_.is_layout_sensitive()) {
+    return nullptr;
+  }
+
+  // Canonicalize dot(<constant>, rhs) to dot(rhs, <constant>) to make the
+  // remainder of this function easier.
+  auto dnums = dot->dot_dimension_numbers();
+  auto lhs_contracting_dims = dnums.lhs_contracting_dimensions();
+  auto rhs_contracting_dims = dnums.rhs_contracting_dimensions();
+  auto* lhs = dot->mutable_operand(0);
+  auto* rhs = dot->mutable_operand(1);
+  if (dot->operand(0)->IsConstant()) {
+    std::swap(lhs, rhs);
+    std::swap(lhs_contracting_dims, rhs_contracting_dims);
+  }
+
+  // Require single contracting dim to make the implementation easier to
+  // track contracting dims.
+  if (dnums.lhs_contracting_dimensions_size() != 1) {
+    return nullptr;
+  }
+
+  // Pattern match Dot(reshape(transpose(input), constant))
+  HloInstruction* reshape;
+  HloInstruction* transpose;
+  HloInstruction* input;
+  HloInstruction* constant;
+  if (!Match(lhs,
+             m::Reshape(&reshape, m::Transpose(&transpose, m::Op(&input)))) ||
+      !Match(rhs, m::Constant(&constant))) {
+    return nullptr;
+  }
+
+  // Check if reshape squishes some dims into one dim, and that if this one
+  // dim is the dot's lhs contracting dim.
+  CHECK(lhs_contracting_dims.size() == 1);
+  const std::vector<std::pair<int64, int64>>& unmodified_dims =
+      ShapeUtil::DimensionsUnmodifiedByReshape(lhs->operand(0)->shape(),
+                                               lhs->shape());
+  if (unmodified_dims.size() != lhs->shape().rank() - 1) {
+    return nullptr;
+  }
+  for (const auto &pair : unmodified_dims) {
+    if (pair.second == lhs_contracting_dims[0]) {
+      // If the contracting dim is unmodified, and only one dim is modified,
+      // then that dim must be a non-contracting dim.
+      return nullptr;
+    }
+  }
+  // Virtually pull the reshape into the dot. Now the dot is equivalent to a
+  // new dot with "unsquished" lhs contracting dims. We don't need to actually
+  // create a new dot instruction. We can just keep track of lhs and
+  // lhs_contracting_dims.
+  CHECK(lhs->operand(0)->shape().rank() > lhs->shape().rank());
+  for (int64 i = 1; i <= lhs->operand(0)->shape().rank() - lhs->shape().rank();
+       ++i) {
+    lhs_contracting_dims.Add(lhs_contracting_dims[0] + i);
+  }
+  lhs = lhs->mutable_operand(0);
+
+  // Check if lhs_transpose only permutes the contracting dims.
+  const auto& transpose_dims = lhs->dimensions();
+  for (int64 i = 0; i < transpose_dims.size(); ++i) {
+    if (transpose_dims[i] != i &&
+        !(absl::c_linear_search(lhs_contracting_dims, i) &&
+          absl::c_linear_search(lhs_contracting_dims, transpose_dims[i]))) {
```

Comment:
Is this second condition necessary?

Reply:
Right. I think it is not necessary. My thought was that if a transpose moves dim i to dim j, then it is natural to check that both i and j should be in contracting dims. For a transpose that moves a contracting dim i to a non-contracting dim j, If we just check i in contracting dims, the check would pass, however there must exist another move that moves a non-contracting dim j' to a contracting dim i' (i' not necessary equal to i), and the check will fail for this j'->i' move. I'd be inclined to keep the second condition for simple logic. What do you think?

Reply:
I think readers will be inclined to think that any conditions inside `if` statements are probably there for a reason, and so will find this confusing.

If we have a condition that we believe must always be true but does not affect the control flow of the program, then canonically we would use a `CHECK`.